### PR TITLE
Skip reverse

### PR DESCRIPTION
**Pull request by remi-dupre**
Tries sparing part of the cost of reverse in fafnir by querying the ES for already existing address associated with a POI.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -66,6 +68,7 @@ pub fn load_and_index_pois(
     let es = args.es.clone();
     let langs = &args.langs;
     let rubber = &mut mimir::rubber::Rubber::new(&es);
+    let poi_dataset = format!("{}_poi_default", &args.dataset); // TODO: is `{}_poi_default` right?
```
Comment (amatissart) [inline src/lib.rs:71]: the current index name format used by mimir is `munin_<type>_<dataset>`, 
and there is potentially 2 different datasets have been imported by Fafnir, defined by `dataset` and  `dataset-nosearch` arguments.

File: src/addresses.rs [hunk 2]
```diff
@@ -95,18 +146,26 @@ pub fn find_address(
...
-            .ok()
-            .and_then(|addrs| addrs.into_iter().next())
-            .map(|addr| {
-                addr.address()
-                    .expect("get_address returned a non-address object")
-            }),
+        _ => {
+            if let Some(addr) = get_current_addr(rubber, poi_dataset, &poi.id) {
+                // Use the address already in ES
+                Some(addr)
```
Comment (amatissart) [inline src/addresses.rs:152]: a few more checks will be necessary here:
 * has the poi position not changed ?
 * has this addr not been read from tags (that are deleted now) ?

File: src/lib.rs [hunk 3]
```diff
@@ -66,6 +70,18 @@ pub fn load_and_index_pois(
     let es = args.es.clone();
     let langs = &args.langs;
     let rubber = &mut mimir::rubber::Rubber::new(&es);
+
+    let poi_creation_date = get_index_creation_date(rubber, &format!("{}_poi", &args.dataset));
+    let addr_creation_date = get_index_creation_date(rubber, &format!("{}_addr", &args.dataset));
```
Comment (amatissart) [inline src/lib.rs:75]: A similar problem here with index names.
Also note that mimir maintains a default alias for each type, so the dataset name is not required to fetch the state of the addr index.
```suggestion
    let poi_creation_date = get_index_creation_date(rubber, &format!("{}_poi_{}", PREFIX, &args.dataset));
    let addr_creation_date = get_index_creation_date(rubber, &format!("{}_addr", PREFIX));
```

File: src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+use mimir::rubber::Rubber;
+
+/// Get creation date of an index as a timestamp.
+pub fn get_index_creation_date(rubber: &mut Rubber, dataset: &str) -> Option<u64> {
```
Comment (amatissart) [inline src/utils.rs:4]: `index` would be more correct than `dataset` for this parameter

File: src/lib.rs (hunk 3 above)
Comment (remi-dupre) [reply src/lib.rs]: It appears that the dataset name is not required either for POIs?

```suggestion
    let poi_creation_date = get_index_creation_date(rubber, &format!("{}_poi", PREFIX));
    let addr_creation_date = get_index_creation_date(rubber, &format!("{}_addr", PREFIX));
```

Querying `https://${ES_ADDR}/_cat/indices/munin_poi?h=creation.date` I actually get expected result \o/

Other issues should also be addressed in f2700ef

File: src/addresses.rs [hunk 5]
```diff
@@ -4,9 +4,49 @@ use mimirsbrunn::admin_geofinder::AdminGeoFinder;
...
+    rubber
+        .get(&query)
+        .map_err(|err| warn!("failed to connect to ES: {:?}", err))
+        .ok()
+        .and_then(|mut res| {
+            res.json()
+                .map_err(|err| warn!("failed to parse ES response: {:?}", err))
+                .ok()
+        })
+        .map(|poi_json: FetchPOI| (poi_json.coord, poi_json.address))
```
Comment (amatissart) [inline src/addresses.rs:47]: How is an unknown POI (404 response) handled here ?
A `warn!` would not be very appropriate, as this is expected for all new POIs.

Comment (remi-dupre) [reply src/addresses.rs]: That would indeed be an irrelevant spam of the logs, it should be fine now: https://github.com/QwantResearch/fafnir/pull/47/files#diff-9b894eb37fb1dd6f240a692c6167c7dfR42-R45

Also, do you think that there would be a clean way to remove these messages:
```
Apr 10 17:29:30.541 INFO Doing post_body_op on /munin_addr,munin_street/_search?ignore_unavailable=true
```
They appear to be raised here: https://github.com/benashford/rs-es/blob/master/src/lib.rs#L152. A line of logs for each request seems to be a bit to much? :/

Comment (amatissart) [reply src/addresses.rs]: I think this message comes from `rs-es`.  
In [our deployed pipeline](https://github.com/QwantResearch/docker_mimir/blob/master/docker-compose.yml#L44) we are using `RUST_LOG=warn,fafnir=info,mimir=info` to hide verbose messages from dependencies.

File: src/addresses.rs [hunk 6]
```diff
@@ -136,18 +208,37 @@ pub fn find_address(
...
-                addr.address()
-                    .expect("get_address returned a non-address object")
-            }),
+        _ => {
+            if !addr_updated {
+                // Fetch the address already attached to the POI to avoid computing an unnecessary
+                // reverse.
+                match get_current_addr(rubber, poi_index, &poi.id) {
+                    CurPoiAddress::None => {}
+                    CurPoiAddress::NotFound => return None,
```
Comment (amatissart) [inline src/addresses.rs:217]: Is that not supposed to be the opposite ? `NotFound` corresponds to POIs that don't exist in the previous index and for which a reverse is always necessary.

Also, the comparison of coords should also be checked when the address is `None`. In this case too the position may have changed, and a reverse could be necessary.

Comment (remi-dupre) [reply src/addresses.rs]: That was right, I just fixed this.

Maybe I should add some tests about this, as it's already pretty hard to do manually.

File: src/addresses.rs [hunk 7]
```diff
@@ -4,9 +4,76 @@ use mimirsbrunn::admin_geofinder::AdminGeoFinder;
...
 use std::ops::Deref;
 use std::sync::Arc;
 
+/// Check if a mimir address originates from OSM data.
+pub fn is_osm_addr(addr: &mimir::Address) -> bool {
+    match addr {
+        mimir::Address::Addr(addr) => addr.id.contains(":osm:"),
+        mimir::Address::Street(st) => st.id.contains(":osm:"),
+    }
+}
```
Comment (amatissart) [inline src/addresses.rs:18]: Argh, we lost a check somewhere. This function is no longer used.

Comment (remi-dupre) [reply src/addresses.rs:18]: Grmf, how did I miss this ... It is supposed to be there: https://github.com/QwantResearch/fafnir/pull/47/files#diff-9b894eb37fb1dd6f240a692c6167c7dfR223

**Pull request merged**